api/types/events/oneof.go

```diff
@@ -401,6 +405,8 @@ func FromOneOf(in OneOf) (AuditEvent, error) {
 		return e, nil
 	} else if e := in.GetWindowsDesktopSessionEnd(); e != nil {
 		return e, nil
+	} else if e := in.GetSessionConnect(); e != nil {
```

Comment:
Shall we refactor this abomination into a type switch, instead of who-knows-how-many sequential method calls?

Courtesy of some creative multi-line editing:

```go
	if in.Event == nil {
		return nil, trace.BadParameter("failed to parse event, session record is corrupted")
	}

	var auditEvent AuditEvent
	switch e := in.Event.(type) {
	case *OneOf_UserLogin:
		auditEvent = e.UserLogin
	case *OneOf_UserCreate:
		auditEvent = e.UserCreate
	case *OneOf_UserDelete:
		auditEvent = e.UserDelete
	case *OneOf_UserPasswordChange:
		auditEvent = e.UserPasswordChange
	case *OneOf_SessionStart:
		auditEvent = e.SessionStart
	case *OneOf_SessionJoin:
		auditEvent = e.SessionJoin
	case *OneOf_SessionPrint:
		auditEvent = e.SessionPrint
	case *OneOf_SessionReject:
		auditEvent = e.SessionReject
	case *OneOf_Resize:
		auditEvent = e.Resize
	case *OneOf_SessionEnd:
		auditEvent = e.SessionEnd
	case *OneOf_SessionCommand:
		auditEvent = e.SessionCommand
	case *OneOf_SessionDisk:
		auditEvent = e.SessionDisk
	case *OneOf_SessionNetwork:
		auditEvent = e.SessionNetwork
	case *OneOf_SessionData:
		auditEvent = e.SessionData
	case *OneOf_SessionLeave:
		auditEvent = e.SessionLeave
	case *OneOf_PortForward:
		auditEvent = e.PortForward
	case *OneOf_X11Forward:
		auditEvent = e.X11Forward
	case *OneOf_SCP:
		auditEvent = e.SCP
	case *OneOf_Exec:
		auditEvent = e.Exec
	case *OneOf_Subsystem:
		auditEvent = e.Subsystem
	case *OneOf_ClientDisconnect:
		auditEvent = e.ClientDisconnect
	case *OneOf_AuthAttempt:
		auditEvent = e.AuthAttempt
	case *OneOf_AccessRequestCreate:
		auditEvent = e.AccessRequestCreate
	case *OneOf_UserTokenCreate:
		auditEvent = e.UserTokenCreate
	case *OneOf_RoleCreate:
		auditEvent = e.RoleCreate
	case *OneOf_RoleDelete:
		auditEvent = e.RoleDelete
	case *OneOf_TrustedClusterCreate:
		auditEvent = e.TrustedClusterCreate
	case *OneOf_TrustedClusterDelete:
		auditEvent = e.TrustedClusterDelete
	case *OneOf_TrustedClusterTokenCreate:
		auditEvent = e.TrustedClusterTokenCreate
	case *OneOf_GithubConnectorCreate:
		auditEvent = e.GithubConnectorCreate
	case *OneOf_GithubConnectorDelete:
		auditEvent = e.GithubConnectorDelete
	case *OneOf_OIDCConnectorCreate:
		auditEvent = e.OIDCConnectorCreate
	case *OneOf_OIDCConnectorDelete:
		auditEvent = e.OIDCConnectorDelete
	case *OneOf_SAMLConnectorCreate:
		auditEvent = e.SAMLConnectorCreate
	case *OneOf_SAMLConnectorDelete:
		auditEvent = e.SAMLConnectorDelete
	case *OneOf_KubeRequest:
		auditEvent = e.KubeRequest
	case *OneOf_AppSessionStart:
		auditEvent = e.AppSessionStart
	case *OneOf_AppSessionChunk:
		auditEvent = e.AppSessionChunk
	case *OneOf_AppSessionRequest:
		auditEvent = e.AppSessionRequest
	case *OneOf_AppCreate:
		auditEvent = e.AppCreate
	case *OneOf_AppUpdate:
		auditEvent = e.AppUpdate
	case *OneOf_AppDelete:
		auditEvent = e.AppDelete
	case *OneOf_DatabaseCreate:
		auditEvent = e.DatabaseCreate
	case *OneOf_DatabaseUpdate:
		auditEvent = e.DatabaseUpdate
	case *OneOf_DatabaseDelete:
		auditEvent = e.DatabaseDelete
	case *OneOf_DatabaseSessionStart:
		auditEvent = e.DatabaseSessionStart
	case *OneOf_DatabaseSessionEnd:
		auditEvent = e.DatabaseSessionEnd
	case *OneOf_DatabaseSessionQuery:
		auditEvent = e.DatabaseSessionQuery
	case *OneOf_SessionUpload:
		auditEvent = e.SessionUpload
	case *OneOf_MFADeviceAdd:
		auditEvent = e.MFADeviceAdd
	case *OneOf_MFADeviceDelete:
		auditEvent = e.MFADeviceDelete
	case *OneOf_BillingCardCreate:
		auditEvent = e.BillingCardCreate
	case *OneOf_BillingCardDelete:
		auditEvent = e.BillingCardDelete
	case *OneOf_LockCreate:
		auditEvent = e.LockCreate
	case *OneOf_LockDelete:
		auditEvent = e.LockDelete
	case *OneOf_BillingInformationUpdate:
		auditEvent = e.BillingInformationUpdate
	case *OneOf_RecoveryCodeGenerate:
		auditEvent = e.RecoveryCodeGenerate
	case *OneOf_RecoveryCodeUsed:
		auditEvent = e.RecoveryCodeUsed
	case *OneOf_WindowsDesktopSessionStart:
		auditEvent = e.WindowsDesktopSessionStart
	case *OneOf_WindowsDesktopSessionEnd:
		auditEvent = e.WindowsDesktopSessionEnd
	case *OneOf_SessionConnect:
		auditEvent = e.SessionConnect
	}
	if auditEvent == nil {
		return nil, trace.BadParameter("received unsupported event %T", in.Event)
	}
```

Reply:
There is also a reflect-based version:

```go
func FromOneOf(in OneOf) (AuditEvent, error) {
	e := in.GetEvent()
	if e == nil {
		return nil, trace.BadParameter("failed to parse event, session record is corrupted")
	}

	// We go from e (isOneOf_Event) -> reflect.Value (*OneOf_SomeStruct) -> reflect.Value(OneOf_SomeStruct).
	elem := reflect.ValueOf(in.GetEvent()).Elem()

	// OneOfs only have one inner field, verify and then read it.
	if elem.NumField() != 1 {
		// This should never happen for proto one-ofs.
		return nil, trace.BadParameter("unexpect number in value %v: %v != 1", elem.Kind(), elem.NumField())
	}

	auditEvent, ok := elem.Field(0).Interface().(AuditEvent)
	if !ok || reflect.ValueOf(auditEvent).IsNil() {
		return nil, trace.BadParameter("received unsupported event %T", in.Event)
	}
	return auditEvent, nil
}
```

Usually I'd steer away from reflect, but this method is a bit ridiculous.

Reply:
@fspmarshall

Reply:
I'm a fan of the reflect-based solution.  I usually hate using reflect, but this switch statement is such a pain.  I think its worth it here.